proposals/record-structs.md

```diff
@@ -0,0 +1,308 @@
+# Record structs
+
+The syntax for a record struct is as follows:
+
+```antlr
+struct_record_declaration
+    : attributes? struct_modifier* 'partial'? 'record' 'struct' identifier type_parameter_list?
+      parameter_list? struct_interfaces? type_parameter_constraints_clause* record_struct_body
+    ;
+
+record_struct_body
+    : struct_body
+    | ';'
+    ;
+```
+
+Record struct types are value types, like other struct types. They implicitly inherit from the class `System.ValueType`.
+The modifiers and members of a record struct are subject to the same restrictions as those of structs
+(accessibility on type, modifiers on members, parameterless instance constructors,
+`base(...)` instance constructor initializers, definite assignment for `this` in constructor, destructors, ...).
+
+See https://github.com/dotnet/csharplang/blob/master/spec/structs.md
+
+But instance field declarations for a record struct are permitted to include variable initializers when there is a primary constructor.
+(TODO: need to confirm, since default values still possible)
+
+Record structs cannot use `ref` modifier.
+
+At most one partial type declaration of a partial record struct may provide a `parameter_list`.
+The `parameter_list` may not be empty.
+
+Record struct parameters cannot use `ref`, `out` or `this` modifiers (but `in` and `params` are allowed).
```

Comment:
Fields cannot have pointer types (that would break our method of comparison, requiring special handling).
"It is an error for an instance field of a record struct to have an unsafe type."

---
In reply to: [557613324](https://github.com/dotnet/csharplang/pull/4307#discussion_r557613324) [](ancestors = 557613324)

Reply:
That's covered in the "properties" section. We grab the name and type (`in` is not part of the type).

"For each record struct parameter of a record struct declaration there is a corresponding public property member whose name and type are taken from the value parameter declaration."

---
In reply to: [557615389](https://github.com/dotnet/csharplang/pull/4307#discussion_r557615389) [](ancestors = 557615389)